crates/cargo-test-support/src/containers.rs

```diff
@@ -0,0 +1,285 @@
+            .as_object()
+            .unwrap()
+            .iter()
+            .map(|(key, value)| {
+                let key = key
+                    .strip_suffix("/tcp")
+                    .expect("expected TCP only ports")
+                    .parse()
+                    .unwrap();
+                let values = value.as_array().unwrap();
+                let value = values
+                    .iter()
+                    .find(|value| value["HostIp"].as_str().unwrap() == "0.0.0.0")
+                    .expect("expected localhost IP");
+                let host_port = value["HostPort"].as_str().unwrap().parse().unwrap();
+                (key, host_port)
+            })
+            .collect()
+    }
+
+    fn wait_till_ready(&self, port_mappings: &HashMap<u16, u16>) {
+        for port in port_mappings.values() {
+            let mut ok = false;
+            for _ in 0..30 {
+                match std::net::TcpStream::connect(format!("127.0.0.1:{port}")) {
+                    Ok(_) => {
+                        ok = true;
+                        break;
+                    }
+                    Err(e) => {
+                        if e.kind() != std::io::ErrorKind::ConnectionRefused {
+                            panic!("unexpected localhost connection error: {e:?}");
+                        }
+                        std::thread::sleep(std::time::Duration::new(1, 0));
+                    }
+                }
+            }
+            if !ok {
+                panic!("no listener on localhost port {port}");
+            }
+        }
+    }
+}
+
+impl ContainerHandle {
+    /// Executes a program inside a running container.
+    pub fn exec(&self, args: &[&str]) -> std::process::Output {
+        ProcessBuilder::new("docker")
+            .args(&["container", "exec", &self.name])
+            .args(args)
+            .exec_with_output()
+            .unwrap()
+    }
+
+    /// Returns the contents of a file inside the container.
+    pub fn read_file(&self, path: &str) -> String {
+        let output = ProcessBuilder::new("docker")
+            .args(&["cp", &format!("{}:{}", self.name, path), "-"])
+            .exec_with_output()
+            .unwrap();
+        let mut ar = tar::Archive::new(output.stdout.as_slice());
+        let mut entry = ar.entries().unwrap().next().unwrap().unwrap();
+        let mut contents = String::new();
+        entry.read_to_string(&mut contents).unwrap();
+        contents
+    }
+}
+
+impl Drop for ContainerHandle {
+    fn drop(&mut self) {
+        // To help with debugging, this will keep the container alive.
+        if std::env::var_os("CARGO_CONTAINER_TEST_KEEP").is_some() {
+            return;
+        }
+        remove_if_exists(&self.name);
+    }
+}
+
+fn remove_if_exists(name: &str) {
+    if let Err(e) = Command::new("docker")
+        .args(&["container", "rm", "--force", name])
+        .output()
+    {
+        panic!("failed to run docker: {e}");
+    }
+}
```

Comment:
While this is test code, I thought panic during `Drop` is problematic.

Reply:
Yea, it should generally always be avoided. In this particular case, I wasn't too concerned because the panic should be extremely unlikely. The only way for it to happen in the drop is if it fails to launch `docker`, but that should only happen if it isn't in PATH (which can't happen because the handle won't be created otherwise), or if the system is running out of memory or file descriptors (in which case you've probably got bigger problems). I was more interested in sharing this function in the two places it is used and to keep the code simple. I can make it return a Result and move the panic to the other location if you think it would make that clearer?

Reply:
Meh.  I leave it to you.  I'll merge as-is and we can always iterate